File: src/cpp/src/core/symmetry.cpp
#include "c4zero/core/symmetry.hpp"

namespace c4zero::core {
namespace {

Coordinates map_xy(int x, int y, Symmetry symmetry) {
  switch (symmetry) {
    case Symmetry::Identity:
      return Coordinates{x, y, 0};
    case Symmetry::Rot90:
      return Coordinates{3 - y, x, 0};
    case Symmetry::Rot180:
      return Coordinates{3 - x, 3 - y, 0};
    case Symmetry::Rot270:
      return Coordinates{y, 3 - x, 0};
    case Symmetry::MirrorX:
      return Coordinates{3 - x, y, 0};
    case Symmetry::MirrorY:
      return Coordinates{x, 3 - y, 0};
    case Symmetry::Diagonal:
      return Coordinates{y, x, 0};
    case Symmetry::AntiDiagonal:
      return Coordinates{3 - y, 3 - x, 0};
  }
  return Coordinates{x, y, 0};
}

Bitboard transform_bits(Bitboard bits, Symmetry symmetry) {
  Bitboard out = 0;
  for (int z = 0; z < kBoardSize; ++z) {
    for (int y = 0; y < kBoardSize; ++y) {
      for (int x = 0; x < kBoardSize; ++x) {
        const Bitboard mask = cell_mask(x, y, z);
        if ((bits & mask) == 0) {
          continue;
        }
        const Coordinates mapped = map_xy(x, y, symmetry);
        out |= cell_mask(mapped.x, mapped.y, z);
      }
    }
  }
  return out;
}

}  // namespace

std::array<Action, kNumActions> action_permutation(Symmetry symmetry) {
  std::array<Action, kNumActions> permutation{};
  for (Action action = 0; action < kNumActions; ++action) {
    const Coordinates xy = action_to_xy(action);
    const Coordinates mapped = map_xy(xy.x, xy.y, symmetry);
    permutation[action] = xy_to_action(mapped.x, mapped.y);
  }
  return permutation;
}

Position transform(const Position& position, Symmetry symmetry) {
  Position out;
  out.current = transform_bits(position.current, symmetry);
  out.opponent = transform_bits(position.opponent, symmetry);
  out.ply = position.ply;
  const Bitboard occupancy = out.occupancy();
  for (Action action = 0; action < kNumActions; ++action) {
    std::uint8_t height = 0;
    while (height < kBoardSize && (occupancy & cell_mask(action, height)) != 0) {
      ++height;
    }
    out.heights[action] = height;
  }
  return out;
}

std::array<float, kNumActions> transform_policy(const std::array<float, kNumActions>& policy, Symmetry symmetry) {
  std::array<float, kNumActions> out{};
  const auto permutation = action_permutation(symmetry);
  for (Action action = 0; action < kNumActions; ++action) {
    out[permutation[action]] = policy[action];
  }
  return out;
}

std::array<std::uint32_t, kNumActions> transform_visits(const std::array<std::uint32_t, kNumActions>& visits, Symmetry symmetry) {
  std::array<std::uint32_t, kNumActions> out{};
  const auto permutation = action_permutation(symmetry);
  for (Action action = 0; action < kNumActions; ++action) {
    out[permutation[action]] = visits[action];
  }
  return out;
}

std::uint16_t transform_legal_mask(std::uint16_t mask, Symmetry symmetry) {
  std::uint16_t out = 0;
  const auto permutation = action_permutation(symmetry);
  for (Action action = 0; action < kNumActions; ++action) {
    if ((mask & (1u << action)) != 0) {
      out |= static_cast<std::uint16_t>(1u << permutation[action]);
    }
  }
  return out;
}

}  // namespace c4zero::core

File: src/python/c4zero_train/symmetry.py
from __future__ import annotations

from enum import IntEnum

import numpy as np

from c4zero_tools.datasets import Sample


BOARD_SIZE = 4
NUM_ACTIONS = 16


class Symmetry(IntEnum):
    IDENTITY = 0
    ROT90 = 1
    ROT180 = 2
    ROT270 = 3
    MIRROR_X = 4
    MIRROR_Y = 5
    DIAGONAL = 6
    ANTI_DIAGONAL = 7


def transform_sample(sample: Sample, symmetry: int | Symmetry) -> Sample:
    symmetry = Symmetry(symmetry)
    permutation = action_permutation(symmetry)
    current_bits = transform_bits(sample.current_bits, symmetry)
    opponent_bits = transform_bits(sample.opponent_bits, symmetry)
    policy = transform_policy(sample.policy, permutation)
    visits = transform_visits(sample.visit_counts, permutation)
    legal_mask = transform_legal_mask(sample.legal_mask, permutation)
    action = permutation[sample.action] if 0 <= sample.action < NUM_ACTIONS else sample.action
    return Sample(
        current_bits=current_bits,
        opponent_bits=opponent_bits,
        heights=heights_from_bits(current_bits | opponent_bits),
        ply=sample.ply,
        game_id=sample.game_id,
        legal_mask=legal_mask,
        action=action,
        policy=policy,
        visit_counts=visits,
        value=sample.value,
    )


def action_permutation(symmetry: Symmetry) -> tuple[int, ...]:
    out = [0] * NUM_ACTIONS
    for action in range(NUM_ACTIONS):
        x, y = action % BOARD_SIZE, action // BOARD_SIZE
        mapped_x, mapped_y = map_xy(x, y, symmetry)
        out[action] = mapped_y * BOARD_SIZE + mapped_x
    return tuple(out)


def transform_bits(bits: int, symmetry: Symmetry) -> int:
    out = 0
    for z in range(BOARD_SIZE):
        for y in range(BOARD_SIZE):
            for x in range(BOARD_SIZE):
                source = cell_index(x, y, z)
                if bits & (1 << source):
                    mapped_x, mapped_y = map_xy(x, y, symmetry)
                    out |= 1 << cell_index(mapped_x, mapped_y, z)
    return out


def transform_policy(policy: np.ndarray, permutation: tuple[int, ...]) -> np.ndarray:
    out = np.zeros_like(policy)
    for action, mapped in enumerate(permutation):
        out[mapped] = policy[action]
    return out.astype(np.float32, copy=False)


def transform_visits(visits: np.ndarray, permutation: tuple[int, ...]) -> np.ndarray:
    out = np.zeros_like(visits)
    for action, mapped in enumerate(permutation):
        out[mapped] = visits[action]
    return out.astype(np.uint32, copy=False)


def transform_legal_mask(mask: int, permutation: tuple[int, ...]) -> int:
    out = 0
    for action, mapped in enumerate(permutation):
        if mask & (1 << action):
            out |= 1 << mapped
    return out


def heights_from_bits(occupancy: int) -> tuple[int, ...]:
    heights = []
    for action in range(NUM_ACTIONS):
        height = 0
        while height < BOARD_SIZE and (occupancy & (1 << cell_index(action % BOARD_SIZE, action // BOARD_SIZE, height))):
            height += 1
        heights.append(height)
    return tuple(heights)


def map_xy(x: int, y: int, symmetry: Symmetry) -> tuple[int, int]:
    if symmetry == Symmetry.IDENTITY:
        return x, y
    if symmetry == Symmetry.ROT90:
        return 3 - y, x
    if symmetry == Symmetry.ROT180:
        return 3 - x, 3 - y
    if symmetry == Symmetry.ROT270:
        return y, 3 - x
    if symmetry == Symmetry.MIRROR_X:
        return 3 - x, y
    if symmetry == Symmetry.MIRROR_Y:
        return x, 3 - y
    if symmetry == Symmetry.DIAGONAL:
        return y, x
    if symmetry == Symmetry.ANTI_DIAGONAL:
        return 3 - y, 3 - x
    raise ValueError(f"unknown symmetry: {symmetry}")


def cell_index(x: int, y: int, z: int) -> int:
    return z * 16 + y * 4 + x
